### DISCUSSION TITLE
Podman support - WIP

### DISCUSSION THREAD

**Discussion by phoozle** [basecamp/kamal#1256] [Show and tell]
UPDATE: https://github.com/phoozle/kamal_podman (I've packaged this solution up as a standalone gem)

Howdy

Podman is a daemonless engine for managing [OCI containers](https://opencontainers.org) and out-of-the-box is arguably more secure than Docker. Being daemonless it also has advantages on the management / auditing front of containers.

Here is a great article explaining the differences to Docker - [Podman vs Docker Comparison](https://betterstack.com/community/guides/scaling-docker/podman-vs-docker/#podman-vs-docker-comparison)

[Original feature request for Podman support](https://github.com/basecamp/kamal/issues/61)

Check out my initial effort here:
https://github.com/phoozle/kamal/tree/podman

I am unsure how my approach would fit into the maintainer's vision for Kamal, but for the time being I am just going for a proof-of-concept. It will be interesting to explore further the features that can be leveraged with Podman, as well as its limitations compared to Docker.

```
# config/deploy.yml
container_manager:
  manager: podman
  options:
    ...
```

It's very early days but I have successfully deployed one of my Rails apps with kamal using podman as the container manager!
![Screenshot 2024-11-27 at 2 10 38 pm](https://github.com/user-attachments/assets/f4043a54-a033-48d2-9b4e-bc6d7cf78e12)

I hope this serves as a starting point for further discussion around the integration of Podman in Kamal. I am hoping for some feedback from the maintainers on my approach and what the general Kamal community thinks about Podman.

Cheers

**Comment by wrb2** [comment]
Thank you! I am super grateful for this.

One thing I wonder about - there is docker or podman both on the "client side" (building and pushing the image) and on the "server side" (running the containers and the proxy). Could those be configurable separately so you can e.g. build using podman and deploy to docker or the other way around?

**Reply by phoozle** [reply]
Thanks!

What would the use cases be? 
Are you concerned about having to run multiple virtual machines for Docker and Podman on a Mac?

**Reply by wrb2** [reply]
I've seen "podman on the client, docker on the server" in some places. But now that I'm thinking about it, I can also imagine the opposite scenario of people running Docker Desktop on Mac but using RedHat or some other Linux in which podman is the default.

**Comment by viktorianer** [comment]
I used Podman with a setup that allows running Docker commands as usual. This way, I could use docker run ... or even docker compose up, etc., seamlessly with Podman.

It would be great if Kamal could work this way too.

**Reply by phoozle** [reply]
so like kamal piping docker commands to the app server? Like a passthrough or something?

Podman command line is mostly on-par with Docker but with some limitations

**Reply by viktorianer** [reply]
Yes. I just used Podman locally and had the following in place: https://podman-desktop.io/docs/migrating-from-docker/managing-docker-compatibility. Additionally, I added docker compose changes on top, but they are not needed for Kamal and Docker.

**Comment by phoozle** [comment]
Just an update on this one:
I have moved development into a standalone gem https://github.com/phoozle/kamal_podman

I figure this is going to be the easier way forward to allow people to start leveraging podman for their Kamal deployments as quickly as possible.

I have just started porting over the work I did in my branch above, into this gem.

**Reply by pedrosimao** [reply]
Hey, what is the current state of this project.
I have worked in a company that got harassed by the Docker commercial team, with threats and so on.
They had to move their entire stack to Podman. And after seeing this happening I use Podman every time I can.
What is missing in your project to be production ready?

**Reply by pedrosimao** [reply]
That being said, if you use your own private registry, don't use a Docker Desktop (and don't make more than 10 million in revenue), there is no need to pay any license. It will probably be 99% of the cases.
But supporting podman is still an amazing idea. We should move away from big corps using open source as a trojan horse for their commercial services. They always have bad intentions and will eventually make the free/open-srouce product worse and worse.

**Comment by boring-dragon** [comment]
Is there any update on this?
